clang/lib/DPCT/BarrierFenceSpaceAnalyzer.cpp

```diff
@@ -7,311 +7,570 @@
 
-bool clang::dpct::BarrierFenceSpaceAnalyzer::Visit(clang::MemberExpr *ME) {
+bool clang::dpct::BarrierFenceSpaceAnalyzer::Visit(const MemberExpr *ME) {
   if (ME->getType()->isPointerType() || ME->getType()->isArrayType()) {
+#ifdef __DEBUG_BARRIER_FENCE_SPACE_ANALYZER
+    std::cout << "Return False case D: "
+              << ME->getBeginLoc().printToString(
+                     DpctGlobalInfo::getSourceManager())
+              << std::endl;
+#endif
     return false;
   }
   return true;
 }
-void clang::dpct::BarrierFenceSpaceAnalyzer::PostVisit(clang::MemberExpr *) {}
+void clang::dpct::BarrierFenceSpaceAnalyzer::PostVisit(const MemberExpr *) {}
 bool clang::dpct::BarrierFenceSpaceAnalyzer::Visit(
-    clang::CXXDependentScopeMemberExpr *) {
+    const CXXDependentScopeMemberExpr *CDSME) {
+#ifdef __DEBUG_BARRIER_FENCE_SPACE_ANALYZER
+  std::cout << "Return False case E: "
+            << CDSME->getBeginLoc().printToString(
+                   DpctGlobalInfo::getSourceManager())
+            << std::endl;
+#endif
   return false;
 }
 void clang::dpct::BarrierFenceSpaceAnalyzer::PostVisit(
-    clang::CXXDependentScopeMemberExpr *) {}
+    const CXXDependentScopeMemberExpr *) {}
+
 bool clang::dpct::BarrierFenceSpaceAnalyzer::Visit(
-    clang::CXXConstructExpr *CCE) {
-  auto Ctor = CCE->getConstructor();
-  std::string CtorName = Ctor->getParent()->getQualifiedNameAsString();
-  if (AllowedDeviceFunctions.count(CtorName) && !isUserDefinedDecl(Ctor))
-    return true;
-  return false;
+    const CXXConstructExpr *CCE) {
+  for (const auto &Arg : CCE->arguments())
+    DeviceFunctionCallArgs.insert(Arg);
+  return true;
 }
 void clang::dpct::BarrierFenceSpaceAnalyzer::PostVisit(
-    clang::CXXConstructExpr *) {}
+    const CXXConstructExpr *) {}
 
-bool clang::dpct::BarrierFenceSpaceAnalyzer::traverseFunction(
-    const clang::FunctionDecl *FD) {
-  if (!this->TraverseDecl(const_cast<clang::FunctionDecl *>(FD))) {
-    return false;
+std::set<const clang::DeclRefExpr *>
+clang::dpct::BarrierFenceSpaceAnalyzer::matchAllDRE(const VarDecl *TargetDecl,
+                                                    const Stmt *Range) {
+  std::set<const DeclRefExpr *> Set;
+  if (!TargetDecl || !Range) {
+    return Set;
   }
-  return true;
+  auto DREMatcher = ast_matchers::findAll(
+      ast_matchers::declRefExpr(ast_matchers::isDeclSameAs(TargetDecl))
+          .bind("DRE"));
+  auto MatchedResults =
+      ast_matchers::match(DREMatcher, *Range, DpctGlobalInfo::getContext());
+  for (auto Node : MatchedResults) {
+    if (auto DRE = Node.getNodeAs<DeclRefExpr>("DRE"))
+      Set.insert(DRE);
+  }
+  return Set;
 }
 
-bool isPointerOperationSafe(const clang::Expr *Pointer) {
-  auto P = getNonImplicitCastNonParenExprParentStmt(Pointer);
-  if (!P)
-    return false;
+std::pair<std::set<const clang::DeclRefExpr *>,
+          std::set<const clang::VarDecl *>>
+clang::dpct::BarrierFenceSpaceAnalyzer::isAssignedToAnotherDREOrVD(
```

Comment:
please add some comment for the functions:  1. scenarios covered. 2.high level design rule/logic.

Reply:
done